### ISSUE TITLE
Can't know what address tx is on when blinded

### ISSUE THREAD

**Issue by Kexkey** [ElementsProject/elements#822]
Using 0.18.1.5.

I import a blinded p2sh-segwit address into a watchonly wallet (watching01.dat):

`{"method":"importaddress","params":["AzpnRT1ZSSFAfupU2i7qQnwb4hhmui3Psyc4z7wARGJ5ip4wFWKW9yqRZMKGDE1bUbVtd8kW6Z5W5D7x","",false]}`

When there's a transaction to that address (from another wallet):

`{"method":"sendtoaddress","params":["AzpnRT1ZSSFAfupU2i7qQnwb4hhmui3Psyc4z7wARGJ5ip4wFWKW9yqRZMKGDE1bUbVtd8kW6Z5W5D7x",0.10000000,"","",false,false,1,"UNSET","b2e15d0d7a0c94e4e2ce0fe6e8691b9e451377f6e46e8045a86f7c4b5d4f0f23"]}`
`{"result":"85a5d20384fe88d728d81e1174ade68c472af2e2303d51380fd1527ef282cd61","error":null,"id":null}`

...the watching node gets it correctly:

`2020-02-18T22:28:26Z [watching01.dat] AddToWallet 85a5d20384fe88d728d81e1174ade68c472af2e2303d51380fd1527ef282cd61  new`

Problem: I can't find what address the tx is related to.  I can only find unconfidential addresses...

```
gettransaction 85a5d20384fe88d728d81e1174ade68c472af2e2303d51380fd1527ef282cd61 true
...no mention of any address
```

```
getrawtransaction 85a5d20384fe88d728d81e1174ade68c472af2e2303d51380fd1527ef282cd61 true
  vout.scriptPubKey.addresses: XBtacEZHPBF9kBRm6UrfEv43yZVTncmenq
  vout.scriptPubKey.addresses: XEgpKAfhKutEtoJK5HkaMY9gaP93GnRz77
```

```
getaddressinfo XBtacEZHPBF9kBRm6UrfEv43yZVTncmenq
  "iswatchonly": false,
getaddressinfo XEgpKAfhKutEtoJK5HkaMY9gaP93GnRz77
  "iswatchonly": true,
  "confidential": "AzppvKt1J3Xt3XM22r1DydataEVVnESR8wYWoe2S4kvdM8tzTVFFCFp7d7xq87XmLJiG2Y3SqQxGjpWp",
```

```
listreceivedbyaddress 0 true true
  {
    "address": "XEgpKAfhKutEtoJK5HkaMY9gaP93GnRz77",
    "amount": {
      "bitcoin": 0.00000000
    },
    "confirmations": 0,
    "label": "",
    "txids": [
    ]
  },
```

Question: if I imported the blinded AzpnRT address and it got detected, I assume the node has the ability to tell the blinded address is involved in the tx.  If I have a bunch of imported addresses, how can I know what address a tx is related to?

**Comment by Kexkey** [comment]
The same thing happens on the spending node.

One way to get the blinded address information for a tx is to do a `gettransaction`, extract the `XEgpKAfhKutEtoJK5HkaMY9gaP93GnRz77` address and do a `getaddressinfo` on the spending wallet.  Then we can read the confidential address.

Would it be possible to use the same address format used in the actual transaction, in the `gettransaction` result?

**Comment by stevenroose** [comment]
Can you check if the transaction was actually blinded?

The problem is that a blinded address is nothing more than an unblinded address + some blinding information. So a sender that doesn't want to blind can always send to the unblinded variant. This means that in the wallet, we keep track of the pubkeys and not the addresses specifically.

We probably should, if a TXO is blinded, show the blinded address and if the TXO is not blinded, show the unblinded address. Can you verify & confirm if it actually is a blinded transaction?

**Comment by Kexkey** [comment]
Yes that's what I was thinking, if the output is blinded it would be very convenient, almost essential, to show the blinded address.  If possible, of course.

I used a standard `sendtoaddress` and I have `blindedaddresses=1` in my elements.conf file.  You can see my `sendtoaddress` call in the OP.  Here is the output of `elements-cli -rpcwallet=spending01.dat gettransaction 85a5d20384fe88d728d81e1174ade68c472af2e2303d51380fd1527ef282cd61 true`:

```
{
  "amount": {
    "bitcoin": 0.00000000
  },
  "fee": {
    "bitcoin": -0.00039660
  },
  "confirmations": 0,
  "trusted": true,
  "txid": "85a5d20384fe88d728d81e1174ade68c472af2e2303d51380fd1527ef282cd61",
  "walletconflicts": [
  ],
  "time": 1582064906,
  "timereceived": 1582064906,
  "bip125-replaceable": "no",
  "details": [
    {
      "address": "XEgpKAfhKutEtoJK5HkaMY9gaP93GnRz77",
      "category": "send",
      "amount": -0.10000000,
      "amountblinder": "45285159330774565b8095c7eec9eb15bd61fcdf29565eb5a9bf8fed1929aef0",
      "asset": "b2e15d0d7a0c94e4e2ce0fe6e8691b9e451377f6e46e8045a86f7c4b5d4f0f23",
      "assetblinder": "61d6df79262a0655ae1cdc8b6e976ee61633adc597c79239727f77111d1d79ee",
      "label": "",
      "vout": 1,
      "fee": 0.00039660,
      "abandoned": false
    },
    {
      "address": "XEgpKAfhKutEtoJK5HkaMY9gaP93GnRz77",
      "category": "receive",
      "amount": 0.10000000,
      "amountblinder": "45285159330774565b8095c7eec9eb15bd61fcdf29565eb5a9bf8fed1929aef0",
      "asset": "b2e15d0d7a0c94e4e2ce0fe6e8691b9e451377f6e46e8045a86f7c4b5d4f0f23",
      "assetblinder": "61d6df79262a0655ae1cdc8b6e976ee61633adc597c79239727f77111d1d79ee",
      "label": "",
      "vout": 1
    }
  ],
  "hex": "..."
}
```

**Comment by FrancisPouliot** [comment]
> Can you check if the transaction was actually blinded?
> 

We did, it was indeed blinded 

> We probably should, if a TXO is blinded, show the blinded address and if the TXO is not blinded, show the unblinded address.

ACK

**Comment by instagibbs** [comment]
you cannot recover the blinded address from chain data.

Have you looked at dumping/importing the master blinding key? Or the specific blinding key? 

`importblindingkey`
`importmasterblindingkey` IIRC

**Comment by instagibbs** [comment]
to be more clear, with on-chain data, the blinding pubkey isn't "written anywhere". It is only stored by wallets.

**Comment by instagibbs** [comment]
Ok I read a bit more, you can tell blindness based on non-zero blinding info, and convert back and forth using `getaddressinfo`.

So ergonomics aside, it's all doable now.

**Comment by Kexkey** [comment]
Thanks for the hints!

It would be convenient though that since I imported the blinded address in the watchonly wallet, I could see it when I `gettransaction` instead of seeing the unblinded address.  After all, if the wallet detected that the transaction is related to it, I assume it internally has the data.

At least from the spender wallet, displaying the blinded address in `gettransaction` output when a blinded address was used would be useful.

My current workaround: when my app receives the request to watch a blinded address, it keeps the blinded and unblinded addresses in its DB so it can link to an incoming transaction.

**Comment by instagibbs** [comment]
Without breaking API we can return the blinded version in parallel, if the output was indeed blinded the the known blinding pubkey.
